Bitcoin for SWAP & HTLC (#35)
comit-network/RFCs: RFC-004-SWAP-Bitcoin.md

# Bitcoin Ledger in SWAP Messages

- RFC-Number: 004
- Status: Draft
- Discussion-issue: -
- Created on: 1 Mar. 2019

# Table of Contents
<!-- markdown-toc start -->

- [Bitcoin Ledger in SWAP Messages](#bitcoin-ledger-in-swap-messages)
- [Table of contents](#table-of-contents)
    - [Description](#description)
    - [The Bitcoin Ledger](#the-bitcoin-ledger)
        - [`network`](#network)
    - [The Bitcoin Asset](#the-bitcoin-asset)
        - [`quantity`](#quantity)
    - [Registry extension](#registry-extension)
        - [The Bitcoin Ledger](#the-bitcoin-ledger-1)
            - [Bitcoin Networks](#bitcoin-networks)
        - [The Bitcoin Asset](#the-bitcoin-asset-1)
- [Examples](#examples)

<!-- markdown-toc end -->

## Description

This RFC specifies how the Bitcoin blockchain and its native asset Bitcoin are described in RFC002 SWAP messages.

> Bitcoin refers specifically to the Bitcoin Core  [Bitcoin Core](https://github.com/bitcoin/bitcoin/) and not any blockchains derived from it.

Reviewer: 🔧
```suggestion
Bitcoin refers specifically to [Bitcoin Core](https://github.com/bitcoin/bitcoin/) and not any blockchains derived from it.
```

As required by RFC002, this RFC defines the *ledger* value `bitcoin` and its associated parameters as well as the *asset* `bitcoin`.
This in combination with subsequent RFCs that define concrete execution steps for particular SWAP protocols enable users to execute and negotiate Bitcoin swaps.

## The Bitcoin Ledger

To specify Bitcoin in one of the RFC002 ledger headers (`alpha_ledger` or `beta_ledger`) use the value `bitcoin` with the following parameter:

### `network`

The `network` parameter describes whether you are intending to do a SWAP with *real* Bitcoin on the `mainnet` or are doing a test SWAP on the `testnet` or a shared `regtest` node.
The `network` parameter is mandatory.

All private test networks like [btcd](https://github.com/btcsuite/btcd)'s `simnet` should be specified as `regtest`.

## The Bitcoin Asset

Bitcoin is also the name of the Bitcoin blockchain's native asset.
This RFC only pertains to the native Bitcoin asset.

> Other non-native assets will be defined in subsequent RFCs.

Reviewer: 🔧
```suggestion
Other non-native assets (such as [colored coins](https://en.bitcoin.it/wiki/Colored_Coins)) will be defined in subsequent RFCs.
```

> Although Bitcoin is nominally the native asset, ownership of Bitcoin is really the ability to unlock *unspent transaction outputs* whose value is measured in *satoshi*.

Reviewer: 🔧
```suggestion
Although Bitcoin is nominally the native asset, ownership of Bitcoin is really the ability to unlock *unspent transaction outputs* (UTXOs) whose value is measured in *satoshi*.
```

1 Bitcoin is simply the name for 100,000,000 satoshi.

To specify Bitcoin in one of the RFC002 asset headers (`alpha_asset` or `beta_asset`) use the value `bitcoin` with the following parameter:

### `quantity`

The `quantity` parameter describes the quantity of satoshi that the asset represents.
The `quantity` parameter is mandatory.

> Its value MUST be a `u64` (note that in the JSON encoding a `u64` is encoded as decimal string like `"100000000"`).

Reviewer: 🤔  I'm not an expert on this, but is `u64` only relevant for rust? What's the type in java for example? 
Maybe we should we generalize the description, e.g.:  
```suggestion
Its value MUST fit into an unsigned 64-bit integer (e.g. in rust a (u64)[https://doc.rust-lang.org/std/primitive.u64.html]) - note that in the JSON encoding a `u64` is encoded as decimal string like `"100000000"`.
```

Reviewer 2: `u64` is what we've been calling it (see `u32` in RFC003). We don't have an RFC that defines what a `u64` or a `u32` means (we should).

Author: I think it's clear enough for any developer what `u64` is. Feel free to define it in the register.

Reviewer 2: The thing we have to define is that `u64` and `u256` should always be encoded as decimal strings and that `32` should always be a JSON number.

# Bitcoin Ledger in SWAP Messages

- RFC-Number: 004
- Status: Draft
- Discussion-issue: -
- Created on: 1 Mar. 2019

# Table of Contents
<!-- markdown-toc start -->

- [Description](#description)
- [The Bitcoin Ledger](#the-bitcoin-ledger)
    - [`network`](#network)
- [The Bitcoin Asset](#the-bitcoin-asset)
    - [`quantity`](#quantity)
- [Registry extension](#registry-extension)
    - [The Bitcoin Ledger](#the-bitcoin-ledger-1)
        - [Bitcoin Networks](#bitcoin-networks)
    - [The Bitcoin Asset](#the-bitcoin-asset-1)
- [Examples](#examples)
    - [SWAP Request](#swap-request)
<!-- markdown-toc end -->

## Description

This RFC specifies how the Bitcoin blockchain and its native asset Bitcoin are described in [RFC002](./RFC-002-SWAP.md) SWAP messages.
Bitcoin refers specifically to [Bitcoin Core](https://github.com/bitcoin/bitcoin/) and not any blockchains derived from it.
As required by [RFC002](./RFC-002-SWAP.md), this RFC defines the *ledger* value `bitcoin` and its associated parameters as well as the *asset* `bitcoin`.

> This in combination with subsequent RFCs that define concrete execution steps for particular SWAP protocols enable users to execute and negotiate Bitcoin swaps.

Author: ```suggestion
This in combination with subsequent RFCs that define concrete execution steps for particular swap protocols enable users to negotiate and execute Bitcoin swaps.
```

1. It is how it's referred in previous RFCs
2. to keep chronological order

Reviewer 2: As far as I can tell we refer to `SWAP protocols` no `swap protocols` in previous PRs. Agreed on the other suggestions.

Author: Here we go: https://github.com/comit-network/RFCs/search?q=%22swap+protocol%22&unscoped_q=%22swap+protocol%22

Reviewer 2: one of them uses SWAP protocol in the same sentence. The others are JSON keys. There is only one that uses "swap protocol" to mean "SWAP protocol". I'm going to say that it's the wrong one and this is right.

## The Bitcoin Ledger

To specify Bitcoin in one of the [RFC002](./RFC-002-SWAP.md) ledger headers (`alpha_ledger` or `beta_ledger`) use the value `bitcoin` with the following parameter:

### `network`

> The `network` parameter describes whether you are intending to do a SWAP with *real* Bitcoin on the `mainnet` or are doing a test SWAP on the `testnet` or a shared `regtest` node.

Author: 🙃 
```suggestion
The `network` parameter describes whether you are intending to do a SWAP with Bitcoin that has *real economic value* on the `mainnet` or are doing a test SWAP on the `testnet` or a shared `regtest` node.
```

Reviewer 2: I think everyone knows what *real* Bitcoin means.

The `network` parameter is mandatory.

All private test networks like [btcd](https://github.com/btcsuite/btcd)'s `simnet` should be specified as `regtest`.

## The Bitcoin Asset

Bitcoin is also the name of the Bitcoin blockchain's native asset.
This RFC only contains protocol definitions for the native Bitcoin asset.
Non-native assets like [colored coins](https://en.bitcoin.it/wiki/Colored_Coins) may be defined in subsequent RFCs.

Although Bitcoin is nominally the native asset, ownership of Bitcoin is really the ability to unlock *unspent transaction outputs* (UTXOs) whose value is measured in *satoshi*.

> 1 Bitcoin is simply the name for 100,000,000 satoshi.

Author: Please just use utf8 like normal people
```suggestion
1 Bitcoin is simply the name for 100,000,000 (10⁸) satoshi.
```

# Bitcoin Ledger in SWAP Messages

- RFC-Number: 004
- Status: Draft
- Discussion-issue: -
- Created on: 1 Mar. 2019

# Table of contents

<!-- toc -->

- [Description](#description)
- [Motivation](#motivation)
- [Content](#content)
  - [Ledger definition](#ledger-definition)
    - [Parameters](#parameters)
  - [Asset definition](#asset-definition)
- [Registry extension](#registry-extension)

<!-- tocstop -->

## Description

This RFC introduces Ledger & Asset definitions to be used in SWAP Messages for [Bitcoin Core](https://github.com/bitcoin/bitcoin/) network.

## Motivation

This RFC aims to define the necessary values and parameters to support Bitcoin Core in SWAP messages.
This, added to complimentary RFCs per SWAP protocol, allows the execution of SWAP involving assets defined on the Bitcoin Core Ledger.

## Content

> 

Reviewer 3: I'd like to see a section intro that says something like:

```
In order to create SWAP messages that make use of the `bitcoin` ledger or the `bitcoin` asset, those need to be defined first.
The next sections define how the `bitcoin` ledger and `bitcoin` asset are encoded when used within a SWAP message.
```

### Ledger definition

This RFC's objective is to uniquely identify the Ledger on which the SWAP will operate.
It is a difficult task as soft and hard forks may happen and there is no guarantee that the conditions identifying today are valid tomorrow.

We define Bitcoin Core as the network and consensus whose standard implementation is [bitcoind](https://github.com/bitcoin/bitcoin/).

- Name: Bitcoin

> - Description: Bitcoin Core Network

Reviewer 3: What does this `Description` refer to? I don't think we define it anywhere, do we?

Author: Not `description` but "description", the English word.

Author: What I mean is that it describes the item.

Reviewer 3: > What I mean is that it describes the item.

I got that. I just don't understand why?

If, as you described above, want to refer to the `Ledger` type here without any further explanation, then the `Description` is misplaced IMO because a `Ledger` doesn't have a `Description`. It also seems redundant to because you said exactly that in the previous paragraph.

Reviewer 2: note: I'm removing the `Description` in favour of just describing it in the actual section.

> - `value`: `bitcoin`

Reviewer 3: IMO, it is a bit too implicit to just list `value` here. Which `value` is meant?

Maybe mention where this comes from and how it is related:

```
RFC002 defined the headers `alpha_ledger` and `beta_ledger` which follow the standard header format defined in RFC001.
```

Author: The headers already refer back to Ledger definition. I don't feel the need to re-refer and re-define.

Reviewer 3: It is not about re-defining. It is more about establishing a context for the reader so that it reads nicely and there is not a bullet points list appeary in the middle of the text flow.

Reviewer 3: For example, I really liked the way @LLFourn wrote this RFC: https://github.com/comit-network/RFCs/blob/master/RFC-003-SWAP-basic.md

Reviewer 2: note: I am going for the explicit verbal approach suggested by @thomaseizinger .

#### Parameters

For security reason, to avoid mainnet asset (valuable) to be swapped with test assets (expendable), a network parameter is defined and must be present for the Bitcoin Ledger.

> There are other known test network such as [btcd](https://github.com/btcsuite/btcd) simnet.

Reviewer 3: :wrench: I'd move these two sentences after the table. The reader doesn't yet know what `regtest` is.

Such network should be represented as `regtest` in SWAP.

- Name: Network
- Description: the network on which to operate
- `parameters` key: `network` 

| `parameters` value |`parameters` description                                             |
|:---                |:---                                                                 |
| `regtest`          | Bitcoin Core regtest or other local/private test network            |
| `testnet`          | Bitcoin Core testnet                                                |
| `mainnet`          | Bitcoin Core mainnet                                                |

### Asset definition

Only the native Bitcoin asset is defined in this RFC.
Other non-native assets should be defined in dedicated RFCs.

> Note that the smallest unit (satoshis) is to be used to avoid ambiguity.

Reviewer 3: ```suggestion
All quantities are encoded without a unit.
They are to be interpreted as satoshis to avoid ambiguity in parser implementations.
```

> Because implementations of Json deserialiser may have different opinions on the maximum size of an integer, the amount is stored in a Json String to avoid unexpected truncation.

Reviewer 3: :wrench: Instead of saying "JSON" here, you could phrase this in a more general way:

```suggestion
To avoid issues with lost precision during parsing, the actual amount is encoded as an integer inside a string.
```

| Name           | Description                   | `value`   | `parameters` key | `parameters` value type | `parameters` description |
|:---            |:----                          |:---       |:---              |:---                     |:---                      |

> | Bitcoin        | Native Bitcoin network asset  | `bitcoin` | `quantity`       | integer in Json string  | Amount in satoshi        |

Reviewer 3: There is no such thing as a "JSON" string.

```suggestion
| Bitcoin        | Native Bitcoin network asset  | `bitcoin` | `quantity`       | integer in a string  | Amount in satoshi        |
```

Author: 1. This comes from the registry
2. We are building on BAM! Json
3. A JSON string is a string as defined in JSON, see [Wikipedia](https://en.wikipedia.org/wiki/JSON#Data_types,_syntax_and_example)

Reviewer 3: > We are building on BAM! Json

> ## Registry extension

Reviewer 3: This section is meant to be complete on its own, f.e.:

https://tools.ietf.org/html/rfc7807#section-6

Author: This looks very redundant.

Reviewer 3: Well, the point would be to not include the whole table in the document body.

Reviewer 2: note: I will have to ask @thomaseizinger what he meant by this.

# Bitcoin Ledger in SWAP Messages

- RFC-Number: 004
- Status: Draft
- Discussion-issue: -
- Created on: 1 Mar. 2019

# Table of contents

<!-- toc -->

> - [Description](#description)

Reviewer 2: Rebuild the TOC

- [Motivation](#motivation)
- [Content](#content)
  - [Ledger definition](#ledger-definition)
    - [Parameters](#parameters)
  - [Asset definition](#asset-definition)
- [Registry extension](#registry-extension)

<!-- tocstop -->

## Description

> This RFC specifies the how the Bitcoin blockchain and its native asset Bitcoin can described in RFC002 SWAP messages.

Reviewer 2: Note I made this change without the "terms of"
